core/src/main/java/ai/timefold/solver/core/preview/api/domain/variable/declarative/ShadowSources.java

```diff
@@ -95,4 +95,20 @@
      * @return A non-empty list of variables the supplier method accesses.
      */
     String[] value();
+
+    /**
+     * If non-empty, this is a property on the entity that will be used
+     * to identify the entity's group.
+     * <p>
+     * When the group key is non-null, the shadow variable will only be calculated
+     * for one entity with that group key, and all other entities with that group key will
+     * have their shadows set to that value.
+     * <p>
+     * When the group key is unspecified or null, the entity is not considered to
+     * be part of a group and will not share variable calculations with other
+     * entities.
```

Comment:
Can the group key ever change? The Javadoc should mention that.
And if it can't (which is what I'm assuming), are we detecting it and failing fast?

This feature opens up a treasure trove of subtle corruption bugs. We need to do as much as we can to prevent the user from shooting themselves in the foot with this.